### PR TITLE
[Work in progress] [macOS] Syphon output

### PR DESCRIPTION
**Pull request by s6joui**
Added Syphon framework integration to share video output with other macOS applications such as Resolume that support Syphon.

Syphon output can be enabled via the `textureSharingEnabled` command line option. Example:

`projectMSDL --textureSharingEnabled 1`

This is a work in progress. Currently facing the following issues:

- How to bundle the Syphon objecive-c framework with the app.(current implementations loads it from /Library/Frameworks/ directory)
- Resizing the window is not well supported in this implementation.

### CODE REVIEW

**Comment by kblaschke** [comment]
If there is a way to not check in all the dependencies into the current repo, but find and link them via CMake' `find_package()`, that'd be a great improvement. IMO, 3rd party libs should stay at their original repos, otherwise the codebase will at some point become a maintenance hell (plus, it's very hard for Linux maintainers to use the distro packages in such a case).

File: src/TextureSharing.h
Comment (revmischa) [inline src/TextureSharing.h:31]:
```suggestion
    GLuint _shareFramebuffer, _shareTexture;
```

**Comment by s6joui** [comment]
> If there is a way to not check in all the dependencies into the current repo, but find and link them via CMake' `find_package()`, that'd be a great improvement. IMO, 3rd party libs should stay at their original repos, otherwise the codebase will at some point become a maintenance hell (plus, it's very hard for Linux maintainers to use the distro packages in such a case).

Agreed! I'll try to figure out how to do it.

File: src/CMakeLists.txt
Comment (kblaschke) [inline src/CMakeLists.txt:21]: Did you check that this still compiles on Windows and Linux, as those OSes don't have Obj-C++ support?

Because on my Linux machine, I get an error configuring the project:

```
Cannot get compiler information:
	Compiler exited with error code 1: /usr/bin/c++ -xobjective-c++ -DAUDIO_IMPL_HEADER=\"AudioCaptureImpl_SDL.h\" -DPOCO_ENABLE_CPP11 -DPOCO_ENABLE_CPP14 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DPOCO_STATIC -DPROJECTMSDL_VERSION=\"2.0.0\" -DPROJECTM_STATIC_DEFINE -DXML_DTD -D_DEBUG -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_REENTRANT -D_THREAD_SAFE -D_XOPEN_SOURCE=500 -isystem/home/kai/Dev/install/projectM-latest-shared/include -isystem/home/kai/Dev/install/Poco/static/include -isystem/home/kai/Dev/install/libSDL/64/include -isystem/home/kai/Dev/install/libSDL/64/include/SDL2 -g -std=gnu++14 -fPIE -fopenmp -fpch-preprocess -v -dD -E
	Using built-in specs.
	COLLECT_GCC=/usr/bin/c++
	Target: x86_64-pc-linux-gnu
	Configured with: /var/tmp/portage/sys-devel/gcc-12.3.0/work/gcc-12.3.0/configure --host=x86_64-pc-linux-gnu --build=x86_64-pc-linux-gnu --prefix=/usr --bindir=/usr/x86_64-pc-linux-gnu/gcc-bin/12 --includedir=/usr/lib/gcc/x86_64-pc-linux-gnu/12/include --datadir=/usr/share/gcc-data/x86_64-pc-linux-gnu/12 --mandir=/usr/share/gcc-data/x86_64-pc-linux-gnu/12/man --infodir=/usr/share/gcc-data/x86_64-pc-linux-gnu/12/info --with-gxx-include-dir=/usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12 --with-python-dir=/share/gcc-data/x86_64-pc-linux-gnu/12/python --enable-languages=c,c++,fortran --enable-obsolete --enable-secureplt --disable-werror --with-system-zlib --enable-nls --without-included-gettext --disable-libunwind-exceptions --enable-checking=release --with-bugurl=https://bugs.gentoo.org/ --with-pkgversion='Gentoo 12.3.0 p2' --with-gcc-major-version-only --disable-esp --enable-libstdcxx-time --enable-lto --disable-libstdcxx-pch --enable-shared --enable-threads=posix --enable-__cxa_atexit --enable-clocale=gnu --enable-multilib --with-multilib-list=m32,m64 --disable-fixed-point --enable-targets=all --enable-libgomp --disable-libssp --disable-libada --disable-cet --disable-systemtap --disable-valgrind-annotations --disable-vtable-verify --disable-libvtv --without-zstd --without-isl --enable-default-pie --enable-default-ssp
	Thread model: posix
	Supported LTO compression algorithms: zlib
	gcc version 12.3.0 (Gentoo 12.3.0 p2) 
	c++: error: language objective-c++ not recognized
	c++: error: language objective-c++ not recognized
```

File: src/CMakeLists.txt
Comment (kblaschke) [inline src/CMakeLists.txt:64]: Is glad required for all platforms, even macOS, when not building Syphon support? I'd suggest making All the Syphon-specific code, libraries etc. dependent on a specific `ENABLE_SYPHON` option or similar, then check for all required dependencies and requirements and give the user proper feedback of something's wrong or missing.

Also, use `cmake_dependent_option()` to only show the option to devs if the build is executed on macOS, as it doesn't make sense or works on any other platform.

File: src/ProjectMSDLApplication.cpp
Comment (kblaschke) [inline src/ProjectMSDLApplication.cpp:163]: Same here - only provide the command-line option in builds which actually have Syphon support built-in. for any other build, it would confuse users.

File: src/glad/include/KHR/khrplatform.h
Comment (kblaschke) [inline src/glad/include/KHR/khrplatform.h:1]: Please don't check in 3rd party headers, rather require them on the build system and use CMake's find_package()/find_include() commands to check their availability and get the proper include path. Ideally, add a CMake Find script in the "cmake" dir on the project top-level dir.

File: src/glad/include/glad/glad.h
Comment (kblaschke) [inline src/glad/include/glad/glad.h:1]: Same here with GLAD: Better use a Find script instead of making a copy in the repo which we need to maintain in the future.

File: src/resources/projectMSDL.properties.in
Comment (kblaschke) [inline src/resources/projectMSDL.properties.in:1]: Also make this addition in the config file dependent on actual Syphon support, so we don't add any options here which aren't supported at all.

**Review by kblaschke — CHANGES_REQUESTED** [review]
See my comments, there are still some things to fix before we can merge this, as the current state of the code will break the build for all platforms except macOS, as Syphon support will be a mandatory feature while it should be optional.

My suggestion here would be this:
- If building on macOS (and only then), provide the user with an option to (forcibly) enable Syphon support.
- If building on macOS and the above option is _not_ on (default), still check for the required libraries and headers (Syphon, GLAD and KHRPlatform). If those are available, set the option to ON if it's _not_ been set to OFF by user already.
- If the option is ON, require the Syphon libraries and fail if they aren't available, so the dev can make sure the Syphon code is definitely compiled in.
- In the build system, code and config files, only add the Spyhon-specific parts (linked libraries/code files in CMake, additional includes and code in C++ and the default options in the config file) if Syphon support is being built. In any other case, leave this out, so it won't break builds on unsupported platforms and also if macOS devs don't want it.

If you need advice on how to best do this in CMake and C++, you can just ask here or in Discord, I'll be glad to help!

**Comment by kblaschke** [comment]
@s6joui still working on this? If not, I'd close the PR.
If you've got anything that works and doesn't affect other platforms where Syphon isn't available, you can just open a new PR.

**Comment by kblaschke** [comment]
No activity or replies for everal months now. I'll close the PR due to inactivity.

If you continue working on it and have something you want to merge, please feel free to open a new PR. In the 4.2 release of libprojectM, there will be new API calls to enable render-to-texture instead of always drawing to the default framebuffer, which would immensely help with this.

I'll also create two feature tickets in the bug tracker, both for Syphon and Spout, to keep this in our minds and track any progress plus allowing for discussions.

**Pull request closed without merging**
